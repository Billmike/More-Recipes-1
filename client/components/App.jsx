import React from 'react';
import PropTypes from 'prop-types';
import { Route, Switch, Redirect } from 'react-router-dom';

import HomePage from './homepage';
import CatalogPage from './catalogPage';
import SingleRecipePage from './singleRecipePage';
import UserDashboardPage from './userRecipes';
import NotFoundPage from './notFoundPage';
import { decodeToken } from '../utils/authenticate';
import '../public/style.scss';

const propTypes = {
  location: PropTypes.shape(),
};

const defaultProps = {
  location: undefined,
};

/**
 * App component
 *
 * @returns {any} - Navigate to the matched route
 */
const App = () => (
  <Switch>
    <Route exact path="/" component={HomePage} />
    <Route exact path="/catalog" component={CatalogPage} />
    <Route exact path="/recipes/:recipeId" component={SingleRecipePage} />
    <Route
      path="/users/:userId/dashboard"
      render={props => (decodeToken()
        ? <UserDashboardPage {...props} />
        : <Redirect to={{
          pathname: '/',
          state: {
            from: props.location
          }
        }}
        />)
      }
    />
    <Route component={NotFoundPage} />
  </Switch>
);

App.propTypes = propTypes;
App.defaultProps = defaultProps;

export default App;
