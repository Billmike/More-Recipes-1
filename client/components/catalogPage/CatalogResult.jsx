import React from 'react';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroll-component';

import RecipeCatalogCard from '../common/RecipeCatalogCard';
import Spinner from '../common/Spinner';

const propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.arrayOf(PropTypes.shape()),
    hasMore: PropTypes.bool.isRequired,
  }).isRequired,
  retrieveMoreRecipes: PropTypes.func.isRequired,
};

/**
 * Function representing catalog result component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} React element
 */
const CatalogResult = (props) => {
  const { recipes: { recipes, hasMore }, retrieveMoreRecipes } = props;

  return (
    <div className="row">
      <div className="catalog-page">
        <InfiniteScroll
          next={retrieveMoreRecipes}
          hasMore={hasMore}
          loader={<div className="col s12 center-align">
            <Spinner size="small" />
          </div>
          }
          endMessage={
            <div className="end-message center-align col s12">
              <b>Yay! You have seen it all</b>
            </div>
          }
        >
          {recipes && recipes.map((recipe, index) =>
            (<RecipeCatalogCard
              {...props}
              key={recipe.id}
              index={index}
              recipe={recipe}
            />))
          }
        </InfiniteScroll>
      </div>
    </div>
  );
};

CatalogResult.propTypes = propTypes;

export default CatalogResult;
