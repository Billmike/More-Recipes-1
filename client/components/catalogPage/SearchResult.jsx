import React from 'react';
import PropTypes from 'prop-types';

import RecipeCatalogCard from '../common/RecipeCatalogCard';
import Spinner from '../common/Spinner';

const propTypes = {
  recipes: PropTypes.shape({
    isFetchingRecipes: PropTypes.bool,
    searchResult: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
};

/**
 * Function representing search result component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} React element
 */
const SearchResult = (props) => {
  const { recipes: { isFetchingRecipes, searchResult } } = props;

  return (
    <div>
      {isFetchingRecipes
        ? <div className="center-align">
          <Spinner size="big" />
        </div>
        : <div>
          {searchResult && searchResult.map((recipe, index) =>
            (<RecipeCatalogCard
              {...props}
              key={recipe.id}
              index={index}
              recipe={recipe}
            />))}
        </div>}
    </div>
  );
};

SearchResult.propTypes = propTypes;

export default SearchResult;
