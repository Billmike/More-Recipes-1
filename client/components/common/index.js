import Header from './Header';
import Footer from './Footer';
import RecipeImage from './RecipeImage';

export default {
  Header,
  Footer,
  RecipeImage,
};
