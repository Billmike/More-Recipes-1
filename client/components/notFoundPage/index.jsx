import React from 'react';
import { Link } from 'react-router-dom';

/**
 * Page not found component
 *
 * @returns {object} React element
 */
const NotFoundPage = () => (
  <div className="row">
    <nav>
      <div className="nav-wrapper deep-orange darken-4">
        <Link
          to="/"
          id="logo"
          className="brand-logo hide-on-small-only"
        >
          More-Recipes
        </Link>
      </div>
    </nav>
    <div className="row center-align">
      <div className="not-found">
        <p className="status-number">404</p>
        <p className="page-data">PAGE NOT FOUND</p>
        <i className="material-icons large">error</i>
      </div>
      <Link to="/">Go home</Link>
    </div>
  </div>

);

export default NotFoundPage;
