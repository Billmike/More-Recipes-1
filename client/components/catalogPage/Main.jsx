import React from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import SearchBar from '../common/SearchBar';
import SearchResult from './SearchResult';
import CatalogResult from './CatalogResult';

const propTypes = {
  recipes: PropTypes.shape({
    searchResult: PropTypes.arrayOf(PropTypes.shape()),
  }),
  location: PropTypes.shape({
    search: PropTypes.string,
  }),
  searchedTerm: PropTypes.string,
  showText: PropTypes.bool.isRequired,
};

const defaultProps = {
  recipes: undefined,
  searchedTerm: undefined,
  location: undefined,
};

/**
 * Function representing main component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} React element
 */
const Main = (props) => {
  const { recipes: { searchResult }, location: { search }, showText } = props;
  let { searchedTerm } = props;

  if (search !== '') {
    const parsed = queryString.parse(search);
    searchedTerm = parsed.list;
  }
  return (
    <div>
      <SearchBar {...props} />
      <div className="row recipes-list">
        <div className="center-align">
          {showText &&
            <h5><em>{searchResult.length === 0
              ? `No Search result found for ${searchedTerm}`
              : `Search results for ${searchedTerm}`
            }</em>
            </h5>
          }
        </div>
        <div>
          {showText
            ? <SearchResult {...props} />
            : <CatalogResult {...props} />}
        </div>

      </div>
    </div>
  );
};

Main.propTypes = propTypes;
Main.defaultProps = defaultProps;

export default Main;
