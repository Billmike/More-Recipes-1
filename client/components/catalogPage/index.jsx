import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actionCreators from '../../actions';
import Header from '../common/Header';
import Main from './Main';
import Footer from '../common/Footer';

const propTypes = {
  searchRecipe: PropTypes.func.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      search: PropTypes.string,
      searchTerm: PropTypes.string,
    }),
  }),
  recipes: PropTypes.shape({
    searchResult: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
};

const defaultProps = {
  location: undefined,
};

/**
 * Class representing catalog page
 *
 * @class CatalogPage
 * @extends {Component}
 */
class CatalogPage extends Component {
  /**
   * Creates an instance of CatalogPage.
   *
   * @memberof CatalogPage
   */
  constructor() {
    super();
    this.state = {
      searchBy: 'name',
      searchTerm: '',
      searchedTerm: '',
      showText: false,
    };
  }

  /**
   * Component did mount lifecycle method
   *
   * @returns {any} Any
   * @memberof CatalogPage
   */
  componentDidMount() {
    const { location: { state }, searchRecipe } = this.props;
    window.scrollTo(0, 0);
    if (state) {
      searchRecipe(state.searchBy, state.searchTerm);
    }
  }

  /**
   * Component will receive props lifecycle method
   *
   * @returns {any} Sets state
   * @memberof CatalogPage
   */
  componentWillReceiveProps() {
    if (this.props.recipes.searchResult) {
      this.setState({
        showText: true,
      });
    }
  }

  /**
   * Function to handle change in search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Changes state to the current search term
   * @memberof CatalogPage
   */
  handleSearchChange = (event) => {
    const { target: { value } } = event;

    if (value === 'name') {
      this.setState(() => ({
        searchBy: value
      }));
    } else if (value === 'ingredients') {
      this.setState(() => ({
        searchBy: value,
      }));
    } else {
      this.setState(() => ({
        searchTerm: value
      }));
    }
  }

  /**
   * Function to handle submitting search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Submits the search term
   * @memberof CatalogPage
   */
  handleSubmitSearch = (event) => {
    event.preventDefault();
    const { searchBy, searchTerm } = this.state;
    const { searchRecipe } = this.props;

    this.setState(() => ({
      searchedTerm: searchTerm,
    }));

    searchRecipe(searchBy, searchTerm);
  }

  /**
   * Render method
   *
   * @returns {object} React element
   * @memberof CatalogPage
   */
  render() {
    return (
      <div>
        <header>
          <Header {...this.props} />
        </header>
        <main>
          <Main
            handleSearchChange={this.handleSearchChange}
            handleSubmitSearch={this.handleSubmitSearch}
            {...this.state}
            {...this.props}
          />
        </main>
        <footer>
          <Footer />
        </footer>

      </div>
    );
  }
}

/**
 * Function to map values from state to props
 *
 * @param {any} state - The state values
 *
 * @returns {object} - The mapped props
 */
const mapStateToProps = state => ({
  recipes: state.recipes,
  user: state.user,
});

/**
 * Function to map dispatch to props
 * Action creators are binded to the dispatch function
 *
 * @param {any} dispatch - The store dispatch function
 *
 * @returns {any} The mapped props
 */
const mapDispatchToProps = dispatch => (
  bindActionCreators(actionCreators, dispatch)
);

CatalogPage.propTypes = propTypes;
CatalogPage.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(CatalogPage);
