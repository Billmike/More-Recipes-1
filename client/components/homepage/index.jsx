import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actionCreators from '../../actions';
import Header from '../common/Header';
import Footer from '../common/Footer';
import Main from './Main';

const propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
  retrieveRecipes: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

/**
 * Class representing the home component
 *
 * @class Home
 *
 * @extends {Component}
 */
class Home extends Component {
  /**
   * Creates an instance of Home.
   *
   * @memberof Home
   */
  constructor() {
    super();
    this.state = {
      search: 'title',
      list: '',
    };
  }

  /**
   * Component did mount function
   *
   * @returns {object} The response from the server
   *
   * @memberof Home
   */
  componentDidMount() {
    const { recipes: { recipes } } = this.props;

    if (!recipes.length > 0) {
      this.props.retrieveRecipes();
    }
  }

  /**
   * Function to handle change in search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Changes state to the current search term
   *
   * @memberof Home
   */
  handleSearchChange = (event) => {
    const { target: { value } } = event;

    if (value === 'title') {
      this.setState(() => ({
        search: value
      }));
    } else if (value === 'ingredients') {
      this.setState(() => ({
        search: value,
      }));
    } else {
      this.setState(() => ({
        list: value
      }));
    }
  }

  /**
   * Function to handle submitting search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Submits the search term
   *
   * @memberof Home
   */
  handleSubmitSearch = (event) => {
    event.preventDefault();
    const { search, list } = this.state;

    if (list !== '') {
      this.props.history.push({
        pathname: '/catalog',
        search: `?search=${search}&list=${list}`,
      });
    }
  }

  /**
   * The render function
   *
   * @returns {object} React element
   *
   * @memberof Home
   */
  render() {
    return (
      <div className="page-body">
        <header>
          <Header {...this.props} />
        </header>
        <main>
          <Main
            handleSearchChange={this.handleSearchChange}
            handleSubmitSearch={this.handleSubmitSearch}
            {...this.state}
            {...this.props}
          />
        </main>
        <footer>
          <Footer />
        </footer>
      </div>
    );
  }
}

/**
 * Function to map values from state to props
 *
 * @param {any} state - The state values
 *
 * @returns {object} - The mapped props
 */
const mapStateToProps = state => ({
  recipes: state.recipes,
  user: state.user,
  userRecipes: state.userRecipes,
});

/**
 * Function to map dispatch to props
 * Action creators are binded to the dispatch function
 *
 * @param {any} dispatch - The store dispatch function
 *
 * @returns {any} The mapped props
 */
const mapDispatchToProps = dispatch => (
  bindActionCreators(actionCreators, dispatch)
);

Home.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(Home);
