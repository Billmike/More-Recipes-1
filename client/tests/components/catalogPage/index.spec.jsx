import React from 'react';

import { CatalogPage } from '../../../components/catalogPage';

describe('CatalogPage component', () => {
  let spy;

  afterEach(() => {
    spy.mockClear();
    localStorage.clear();
  });
  it('should render component', () => {
    const props = {
      searchRecipe: jest.fn(),
      location: {
        search: '',
      },
      history: {
        replace: jest.fn(),
      },
      recipes: {
        recipes: [],
        searchPerformed: false,
      },
      retrieveRecipes: jest.fn(),
    };
    spy = jest.spyOn(CatalogPage.prototype, 'componentDidMount');
    const wrapper = shallow(<CatalogPage {...props} />);

    expect(wrapper).toBeDefined();
    expect(spy).toHaveBeenCalled();
    expect(props.retrieveRecipes).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should search for recipe if search is performed', () => {
    const props = {
      searchRecipe: jest.fn(),
      location: {
        search: '?search=title&list=rice',
      },
      history: {
        replace: jest.fn(),
      },
      recipes: {
        recipes: [],
        searchPerformed: false,
      },
      retrieveRecipes: jest.fn(),
    };
    spy = jest.spyOn(CatalogPage.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<CatalogPage {...props} />);

    expect(wrapper).toBeDefined();
    expect(props.searchRecipe).toHaveBeenCalled();

    wrapper.setProps({
      recipes: {
        recipes: [{
          id: 5,
          title: 'Macaroni',
          category: 'Dinner',
          description: 'Tasty macaroni',
          preparationTime: 45,
          ingredients: 'Macaroni, groundnut oil, salt',
          directions: 'Do this, do that',
          recipeImage: 'Recipe image',
          upvotes: 0,
          downvotes: 0,
          views: 1,
          favorites: 0,
        }],
        searchPerformed: true,
      }
    });
    expect(spy).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
  it('should handle instance method events', () => {
    const props = {
      searchRecipe: jest.fn(),
      location: {
        search: '',
      },
      history: {
        replace: jest.fn(),
      },
      recipes: {
        recipes: [],
        searchPerformed: false,
      },
      retrieveRecipes: jest.fn(),
    };
    const wrapper = shallow(<CatalogPage {...props} />);

    expect(wrapper).toBeDefined();

    wrapper.instance().handleSearchChange({ target: { value: 'title' } });
    expect(wrapper.instance().state.search).toEqual('title');
    wrapper.instance().handleSearchChange({ target: { value: 'ingredients' } });
    expect(wrapper.instance().state.search).toEqual('ingredients');
    wrapper.instance().handleSearchChange({ target: { value: 'pepper' } });
    expect(wrapper.instance().state.list).toEqual('pepper');

    wrapper.instance().handleSubmitSearch({ preventDefault: jest.fn() });
    expect(wrapper.instance().state.searchedTerm).toEqual('pepper');
    expect(props.searchRecipe).toBeCalled();

    wrapper.instance().handleSearchCategory({ target: { name: 'Appetizer' } });
    expect(props.searchRecipe).toBeCalled();
    expect(props.history.replace).toBeCalled();

    wrapper.instance().retrieveMoreRecipes();
    expect(props.retrieveRecipes).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
});
