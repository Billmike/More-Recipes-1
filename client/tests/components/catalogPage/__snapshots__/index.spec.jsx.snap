// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogPage component should handle instance method events 1`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "replace": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=ingredients&list=pepper",
                },
              ],
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=category&list=Appetizer",
                },
              ],
            ],
          },
        }
      }
      location={
        Object {
          "search": "",
        }
      }
      recipes={
        Object {
          "recipes": Array [],
          "searchPerformed": false,
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
            Array [
              8,
              8,
            ],
          ],
        }
      }
      searchRecipe={
        [MockFunction] {
          "calls": Array [
            Array [
              "ingredients",
              "pepper",
            ],
            Array [
              "category",
              "Appetizer",
            ],
          ],
        }
      }
    />
  </header>
  <main
    className="grey lighten-3"
  >
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "replace": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=ingredients&list=pepper",
                },
              ],
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=category&list=Appetizer",
                },
              ],
            ],
          },
        }
      }
      limit={8}
      list=""
      location={
        Object {
          "search": "",
        }
      }
      offset={8}
      recipes={
        Object {
          "recipes": Array [],
          "searchPerformed": false,
        }
      }
      retrieveMoreRecipes={[Function]}
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
            Array [
              8,
              8,
            ],
          ],
        }
      }
      scrollToTop={[Function]}
      search="title"
      searchPerformed={false}
      searchRecipe={
        [MockFunction] {
          "calls": Array [
            Array [
              "ingredients",
              "pepper",
            ],
            Array [
              "category",
              "Appetizer",
            ],
          ],
        }
      }
      searchedTerm=""
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;

exports[`CatalogPage component should render component 1`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "replace": [MockFunction],
        }
      }
      location={
        Object {
          "search": "",
        }
      }
      recipes={
        Object {
          "recipes": Array [],
          "searchPerformed": false,
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      searchRecipe={[MockFunction]}
    />
  </header>
  <main
    className="grey lighten-3"
  >
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "replace": [MockFunction],
        }
      }
      limit={8}
      list=""
      location={
        Object {
          "search": "",
        }
      }
      offset={8}
      recipes={
        Object {
          "recipes": Array [],
          "searchPerformed": false,
        }
      }
      retrieveMoreRecipes={[Function]}
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      scrollToTop={[Function]}
      search="title"
      searchPerformed={false}
      searchRecipe={[MockFunction]}
      searchedTerm=""
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;

exports[`CatalogPage component should search for recipe if search is performed 1`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "replace": [MockFunction],
        }
      }
      location={
        Object {
          "search": "?search=title&list=rice",
        }
      }
      recipes={
        Object {
          "recipes": Array [
            Object {
              "category": "Dinner",
              "description": "Tasty macaroni",
              "directions": "Do this, do that",
              "downvotes": 0,
              "favorites": 0,
              "id": 5,
              "ingredients": "Macaroni, groundnut oil, salt",
              "preparationTime": 45,
              "recipeImage": "Recipe image",
              "title": "Macaroni",
              "upvotes": 0,
              "views": 1,
            },
          ],
          "searchPerformed": true,
        }
      }
      retrieveRecipes={[MockFunction]}
      searchRecipe={
        [MockFunction] {
          "calls": Array [
            Array [
              "title",
              "rice",
            ],
          ],
        }
      }
    />
  </header>
  <main
    className="grey lighten-3"
  >
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "replace": [MockFunction],
        }
      }
      limit={8}
      list=""
      location={
        Object {
          "search": "?search=title&list=rice",
        }
      }
      offset={0}
      recipes={
        Object {
          "recipes": Array [
            Object {
              "category": "Dinner",
              "description": "Tasty macaroni",
              "directions": "Do this, do that",
              "downvotes": 0,
              "favorites": 0,
              "id": 5,
              "ingredients": "Macaroni, groundnut oil, salt",
              "preparationTime": 45,
              "recipeImage": "Recipe image",
              "title": "Macaroni",
              "upvotes": 0,
              "views": 1,
            },
          ],
          "searchPerformed": true,
        }
      }
      retrieveMoreRecipes={[Function]}
      retrieveRecipes={[MockFunction]}
      scrollToTop={[Function]}
      search="title"
      searchPerformed={true}
      searchRecipe={
        [MockFunction] {
          "calls": Array [
            Array [
              "title",
              "rice",
            ],
          ],
        }
      }
      searchedTerm=""
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;
